### PR TITLE
Port the zfs driver

### PR DESCRIPTION
**Pull request by stgraber**

### CODE REVIEW

**Comment by stgraber** [comment]
This PR includes https://github.com/lxc/lxd/pull/6686

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

File: test/includes/storage.sh [hunk 1]
```diff
@@ -130,8 +130,8 @@ umount_loops() {
 }
 
 storage_compatible() {
-    if [ "${1}" = "cephfs" ] || [ "${1}" = "dir" ] || [ "${1}" = "btrfs" ]; then
-        if [ "${2}" = "cephfs" ] || [ "${2}" = "dir" ] || [ "${2}" = "btrfs" ]; then
+    if [ "${1}" = "cephfs" ] || [ "${1}" = "dir" ] || [ "${1}" = "btrfs" ] || [ "${1}" = "zfs" ]; then
```
Comment (tomponline) [inline test/includes/storage.sh]: @stgraber should I add lvm to this section too?

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 2]
```diff
@@ -0,0 +1,1395 @@
...
+		revert.Add(func() { d.DeleteVolume(fsVol, op) })
+	}
+
+	// Mount the dataset.
+	_, err := d.MountVolume(vol, op)
+	if err != nil {
+		return err
+	}
+
+	if vol.contentType == ContentTypeFS {
```
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go]: @stgraber looks like this block is a duplicate of the one at https://github.com/lxc/lxd/pull/6688/files#diff-889526b02b65fe324db3c38b4a6fa371R32

File: test/includes/storage.sh (hunk 1 above)
Comment (stgraber) [reply test/includes/storage.sh]: yes, all ported drivers need to be added there

File: lxd/storage/drivers/driver_zfs_volumes.go (hunk 2 above)
Comment (stgraber) [reply lxd/storage/drivers/driver_zfs_volumes.go]: It's not. The first one creates the mountpoint and sets the permissions on the mountpoint.
The second one sets the permission on the mount itself.

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 3]
```diff
@@ -0,0 +1,1395 @@
...
+	// Mount the dataset.
+	_, err := shared.RunCommand("zfs", "mount", d.dataset(vol, false))
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// UnmountVolume simulates unmounting a volume. As dir driver doesn't have volumes to unmount it
```
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go]: @stgraber some mentions of dir driver.

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 4]
```diff
@@ -0,0 +1,1395 @@
...
+
+		if strings.TrimSpace(output) == d.dataset(vol, false) {
+			return entryPath, nil
+		}
+	}
+
+	return "", fmt.Errorf("Could not locate a zvol for %s", d.dataset(vol, false))
+}
+
+// MountVolume simulates mounting a volume. As dir driver doesn't have volumes to mount it returns
```
Comment (tomponline) [inline lxd/storage/drivers/driver_zfs_volumes.go]: dir driver mentioned.

File: lxd/storage/drivers/driver_zfs_volumes.go (hunk 2 above)
Comment (tomponline) [reply lxd/storage/drivers/driver_zfs_volumes.go]: @stgraber ah ok, is that a ZFS specific thing? Wondering if need to do something similar for LVM.

Comment (stgraber) [reply lxd/storage/drivers/driver_zfs_volumes.go]: LVM should do the same, so should CEPH.

**Pull request merged**